### DISCUSSION TITLE
Create a query specifying two atoms are not bonded

### DISCUSSION THREAD

**Discussion by jasondbiggs** [rdkit/rdkit#4501] [Q&A] [answered]
Is it possible to create a query mol that specifies two atoms with queries and specifies that no bond exists between those atoms?  For instance the SMARTS string "[#6].[#7]` will match any pair of carbon and nitrogen atoms.  How could I match only those pairs that don't have a bond to each other?  

I don't think that this is possible through SMARTS, but I am wondering if I could, in principle, make such a query mol in C++?

I actually think this is part of a larger set of queries, where some kind of post-match condition is imposed.  I could imagine a query for a substructure of two atoms a certain topological distance apart from each other (or 3D distance).

**Comment by ptosco** [comment]
@jasondbiggs You could use
```
from rdkit import Chem
mol = Chem.MolFromSmiles("CCNC")
query = Chem.MolFromSmarts("[#6].[#7]")
mol.GetSubstructMatches(query)
((0, 2), (1, 2), (3, 2))
query_not_bonded = Chem.MolFromSmarts("[$([#6].[#7]);!$([#6]~[#7])].[#7]")
mol.GetSubstructMatches(query_not_bonded)
((0, 2),)
```

On the topic of programmatic queries, this is now supported also from Python; check the [`rdkit.Chem.rdqueries`](http://rdkit.org/docs/source/rdkit.Chem.rdqueries.html?highlight=query#module-rdkit.Chem.rdqueries) module.

**Comment by jasondbiggs** [comment]
@ptosco thank you, that is a good suggestion, but not exactly what I mean.  Ideally I could find any two carbon and nitrogen atoms that aren't bonded, but that SMARTS pattern puts the extra condition that the carbon can't be bonded to *any* nitrogen. So it would not find a match in this molecule
<img width="234" alt="image" src="https://user-images.githubusercontent.com/16337306/132954664-0ff6aa30-511a-4c3c-84a8-901292963cda.png">
even though there is one pair of C,N that aren't bonded.

As I see the types of queries I can make in c++, or the rdqueries you pointed to in python, I can make either a query atom, a query bond, or a query to match a particular molecule Property value.  This allows for many types of substructure patterns, but doesn't seem to cover the type of query I am looking to make.

**Comment by ptosco** [comment]
@jasondbiggs Ah, OK, I see. I can't imagine of a SMARTS query alone that can do the job on that molecule, nor there is a programmatic query that would help.
I think this can only be addressed with a post-processing step at the moment, e.g.
```
from rdkit import Chem
mol = Chem.MolFromSmiles("NOCN")
query = Chem.MolFromSmarts("[#6].[#7]")
tuple(m for m in mol.GetSubstructMatches(query) if not mol.GetBondBetweenAtoms(*m))
((2, 0),)
```
You could do a similar post-processing on arbitrary topologic distances using a topological distance matrix as a post-condition.

**Comment by greglandrum** [comment, marked as answer]
This is exactly what the `extraFinalCheck` function in `SubstructMatchParameters` is for.
`extraFinalCheck` is a function which is called with the molecule and a vector of potentially matching atom indices from that molecule. It should return whether or not the match is accepted.
![image](https://user-images.githubusercontent.com/540511/132972321-6867024c-bed2-4869-a3d9-3397427c1e84.png)

Here's the complete notebook demonstrating this:
https://gist.github.com/greglandrum/73eecb38eccb0a5f59605dba53640b35

**Reply by ptosco** [reply]
@greglandrum Thanks for chiming in, Greg! I had seen that function, but I didn’t think of using it, that’s very elegant!

**Reply by jasondbiggs** [reply]
Thanks @greglandrum that works perfect!
